### PR TITLE
[SPARK-32755][SQL] Maintain the order of expressions in AttributeSet and ExpressionSet

### PR DESCRIPTION
**Pull request by dbaliafroozeh**
### What changes were proposed in this pull request?
This PR changes `AttributeSet` and `ExpressionSet` to maintain the insertion order of the elements. More specifically, we:
- change the underlying data structure of `AttributeSet` from `HashSet` to `LinkedHashSet` to maintain the insertion order.
- `ExpressionSet` already uses a list to keep track of the expressions, however, since it is extending Scala's immutable.Set class, operations such as map and flatMap are delegated to the immutable.Set itself. This means that the result of these operations is not an instance of ExpressionSet anymore, rather it's a implementation picked up by the parent class. We also remove this inheritance from `immutable.Set `and implement the needed methods directly. ExpressionSet has a very specific semantics and it does not make sense to extend `immutable.Set` anyway.
- change the `PlanStabilitySuite` to not sort the attributes, to be able to catch changes in the order of expressions in different runs.

### Why are the changes needed?
Expressions identity is based on the `ExprId` which is an auto-incremented number. This means that the same query can yield a query plan with different expression ids in different runs. `AttributeSet` and `ExpressionSet` internally use a `HashSet` as the underlying data structure, and therefore cannot guarantee the a fixed order of operations in different runs. This can be problematic in cases we like to check for plan changes in different runs.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Passes `PlanStabilitySuite` after regenerating the golden files.

### CODE REVIEW

**Comment by hvanhovell** [comment]
ok to test

**Comment by hvanhovell** [comment]
jenkins retest this please

**Review by hvanhovell — APPROVED** [review]
LGTM

File: sql/catalyst/src/main/scala-2.12/org/apache/spark/sql/catalyst/expressions/ExpressionSet.scala [hunk 1]
```diff
@@ -27,6 +27,10 @@ object ExpressionSet {
     expressions.foreach(set.add)
     set
   }
+
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala-2.12/org/apache/spark/sql/catalyst/expressions/ExpressionSet.scala:30]: We should apply the same change in `ExpressionSet` under the `scala-2.13` source tree. @dbaliafroozeh can you open a followup  PR?

Comment (dbaliafroozeh) [reply sql/catalyst/src/main/scala-2.12/org/apache/spark/sql/catalyst/expressions/ExpressionSet.scala:30]: @cloud-fan good catch, I thought I already delete the ExpressionSet in 2.13. Note that we don't want it anymore as ExpressionSet doesn't extend Set anymore. I'll open a followup PR for that.

**Comment by LuciferYang** [comment]
Sorry to leave a message in a completed issue. @cloud-fan @dbaliafroozeh This patch seems to bring about some different behavior between use Scala 2.12 and Scala 2.13.

I found that the number of failed cases increased with this patch of the sub-suites of  `PlanStabilitySuite`.

For example, if we execute 

```
mvn clean test -pl sql/core -Dtest=none -DwildcardSuites=org.apache.spark.sql.TPCDSV2_7_PlanStabilitySuite -Pscala-2.13 -am 
```

The test result with out this patch is 

```
Tests: succeeded 32, failed 0, canceled 0, ignored 0, pending 0
All tests passed.
```

and with this patch is 

```
Tests: succeeded 1, failed 31, canceled 0, ignored 0, pending 0
*** 31 TESTS FAILED ***
```

I haven't found the root cause yet. Do you have any good ideas for fix this problems?

**Comment by cloud-fan** [comment]
This should have been merged before we have `PlanStabilitySuite`, as the query plans in golden files were kind of random previously. That's why this PR updates `PlanStabilitySuite`.

If your spark fork has different golden files for `PlanStabilitySuite`, you should just re-generate golden files after this patch.

**Comment by LuciferYang** [comment]
@cloud-fan Maybe I didn't describe it clearly, now I use the master of spark-source to execute  maven test with Scala 2.12

```
mvn clean test -pl sql/core -Dtest=none -DwildcardSuites=org.apache.spark.sql.TPCDSV2_7_PlanStabilitySuite -am 

```
All tests passed.


execute  maven test with Scala 2.13

```
mvn clean test -pl sql/core -Dtest=none -DwildcardSuites=org.apache.spark.sql.TPCDSV2_7_PlanStabilitySuite -am -Pscala-2.13 

```
31 TESTS FAILED

without this patch execute  maven test , both Scala 2.12 and Scala 2.13 `All tests passed`

**Comment by LuciferYang** [comment]
So always need to re-generate golden files  with Scala 2.13? Or we need to use different golden files for different Scala verision, feels a little unreasonable...

**Comment by cloud-fan** [comment]
Interesting. So `AttributeSet` and `ExpressionSet` behave differently under scala 2.12 and 2.13.  @Ngone51 can you take a look?

**Comment by LuciferYang** [comment]
@Ngone51 need some simple fix on compilation for Scala 2.13 , `QueryPlan` and  `ShuffleBlockFetcherIterator`.

**Comment by cloud-fan** [comment]
@LuciferYang do you have a branch that contains the compilation fix?

**Comment by LuciferYang** [comment]
@cloud-fan @Ngone51  we can use https://github.com/apache/spark/pull/29660

**Comment by LuciferYang** [comment]
@cloud-fan @Ngone51 I think the reason for the problem is 

https://github.com/apache/spark/blob/e7d9a245656655e7bb1df3e04df30eb3cc9e23ad/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/AttributeSet.scala#L106-L113

The `--` of `LinkedHashSet` method is different between Scala 2.12 and Scala 2.13.

The impl of Scala 2.12(`SetLike.scala`) as follow:
```
override def --(xs: GenTraversableOnce[A]): This = clone() --= xs.seq
``` 

The impl of Scala 2.13(`Set.scala`)  as follow:

```
def -- (that: IterableOnce[A]): C = fromSpecific(coll.toSet.removedAll(that))
``` 

From the above code we can found that after `baseSet -- otherSet.baseSet`, the result order changed .

Maybe we can use `baseSet.diff(otherSet.baseSet)` instead of `baseSet -- otherSet.baseSet`, but I need to confirm the feasibility.

**Comment by dbaliafroozeh** [comment]
@cloud-fan @LuciferYang we can also try to use Java's LinkedHashSet here if there is a difference between different versions of Scala's mutable.LinkedHashSet.

**Comment by hvanhovell** [comment]
@dbaliafroozeh `--` returns a set, which gets converted into linked set by `AttributeSet.apply(..)`. The 2.12 implementation will actually return a linked set, whereas the new probably returns a set.

@LuciferYang can you open a PR that basically uses 2.12 implementation inside AttributeSet?

**Comment by LuciferYang** [comment]
> @LuciferYang can you open a PR that basically uses 2.12 implementation inside AttributeSet?

Ok ~ I will give a new followup pr

**Comment by LuciferYang** [comment]
@hvanhovell use `clone() --= xs.seq` instead of `--`?

**Comment by cloud-fan** [comment]
I'd prefer `diff` if it works, as it's simpler.

**Comment by hvanhovell** [comment]
Anything that explicitly maintains the insertion order (i.e. returns a LinkedHashSet) will do :).

**Comment by hvanhovell** [comment]
`diff` returns a set and does not fall in that category.

**Pull request closed without merging**